`zephyr_library_compile_options()` fails to apply if the same setting is set for multiple libraries in a single project

**Describe the bug**
In a project that has multiple Zephyr libraries (created with either the `zephyr_library()` or `zephyr_library_named()` macros), using the `zephyr_library_compile_options()` function to apply the same option to multiple projects will only apply to the first.

I understand it is intended behavior to only apply a given option once to a given library (following CMake's example), but it seems to me that the same option should be able to be applied to multiple different libraries.

My tangible use case of this is if you are using out of tree modules to bring custom drivers into a project. In that case I want to apply different compiler error/warning settings different from Zephyr for my application code as well as my driver code. I set them for the implicit `app` library, but when I try to set them in my driver module, since those same compiler options have already been applied, they are getting silently dropped. (see 'Additional context' below for what I believe to be the root cause and a possible solution).

**To Reproduce**
Steps to reproduce the behavior:
The following code based off the `hello_world` example will reproduce this (zip file with this project is also attached: [project_test.zip](https://github.com/zephyrproject-rtos/zephyr/files/8258598/project_test.zip)).

"CMakeLists.txt"
```cmake
# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.13.1)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(hello_world)

zephyr_library_compile_options(-Werror=unused-variable)

target_sources(app PRIVATE main.c)

# Create a second lib
zephyr_library_named(foo)

# setting this will not get passed to compiling main.c for the foo lib
zephyr_library_compile_options(-Werror=unused-variable)

zephyr_library_sources(main.c)
```

"main.c"
```c
/*
 * Copyright (c) 2012-2014 Wind River Systems, Inc.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <zephyr.h>
#include <sys/printk.h>

static int unused_var;

void main(void)
{
	printk("Hello World! %s\n", CONFIG_BOARD);
}
```

"prj.conf"
```
# nothing here
```
* in a workspace copy those files into a `project_test` folder
* Build with `west build -b native_posix project_test`
* You'll get an error for undefined variable in the main.c build with the app, but not the one for the foo library (just a warning).

**Expected behavior**
I would expect the compiler options to apply to both libraries.

**Impact**
This prevents people from applying the same compiler options to both their app and external modules. This limits the ability to have custom compiler behavior when working with custom drivers.

**Logs and console output**
```sh
[8/92] Building C object CMakeFiles/foo.dir/main.c.obj
/workdir/project_test/main.c:10:12: warning: ‘unused_var’ defined but not used [-Wunused-variable]
   10 | static int unused_var;
      |            ^~~~~~~~~~
[63/92] Building C object CMakeFiles/app.dir/main.c.obj
FAILED: CMakeFiles/app.dir/main.c.obj 
ccache /usr/bin/gcc -DBUILD_VERSION=v2.6.0-rc1-ncs1 -DKERNEL -DUSE_PARTITION_MANAGER=0 -D_FORTIFY_SOURCE=2 -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/posix/inf_clock -I/workdir/zephyr/boards/posix/native_posix -I/workdir/nrf/include -Os -imacros /workdir/build-hello/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -imacros /workdir/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/project_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -include /workdir/zephyr/arch/posix/include/posix_cheats.h -fno-freestanding -std=c11 -Werror=unused-variable -MD -MT CMakeFiles/app.dir/main.c.obj -MF CMakeFiles/app.dir/main.c.obj.d -o CMakeFiles/app.dir/main.c.obj -c /workdir/project_test/main.c
/workdir/project_test/main.c:10:12: error: ‘unused_var’ defined but not used [-Werror=unused-variable]
   10 | static int unused_var;
      |            ^~~~~~~~~~
cc1: some warnings being treated as errors
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Commit SHA or Version used: I'm on an older version through the Nordic SDK (v2.6.0-rc1-ncs1), but from code inspection this appears to still be an issue on `main`

**Additional context**
I think there is a pretty simple solution to this, providing that the above bug is not the intended behavior. In `zephyr_library_compile_options()` a given compiler option is given 'uniqueness' by taking the MD5 sum of it. If we included the library name within the data being hashed for uniqueness, a particular given option could be applied to any number of libraries, but still once and only once per a given library.

Currenly we have
```cmake
string(MD5 uniqueness ${item})
```

I would propose a change to something like
```cmake
string(MD5 uniqueness ${item}_${ZEPHYR_CURRENT_LIBRARY})
```

With that fix (and running with `ninja -K 0` we get the expected output
```sh
FAILED: CMakeFiles/foo.dir/main.c.obj 
ccache /usr/bin/gcc -DBUILD_VERSION=v2.6.0-rc1-ncs1 -DKERNEL -DUSE_PARTITION_MANAGER=0 -D_FORTIFY_SOURCE=2 -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/posix/inf_clock -I/workdir/zephyr/boards/posix/native_posix -I/workdir/nrf/include -Os -imacros /workdir/build-hello/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -imacros /workdir/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/project_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -include /workdir/zephyr/arch/posix/include/posix_cheats.h -fno-freestanding -std=c11 -Werror=unused-variable -MD -MT CMakeFiles/foo.dir/main.c.obj -MF CMakeFiles/foo.dir/main.c.obj.d -o CMakeFiles/foo.dir/main.c.obj -c /workdir/project_test/main.c
/workdir/project_test/main.c:10:12: error: ‘unused_var’ defined but not used [-Werror=unused-variable]
   10 | static int unused_var;
      |            ^~~~~~~~~~
cc1: some warnings being treated as errors
[35/69] Building C object CMakeFiles/app.dir/main.c.obj
FAILED: CMakeFiles/app.dir/main.c.obj 
ccache /usr/bin/gcc -DBUILD_VERSION=v2.6.0-rc1-ncs1 -DKERNEL -DUSE_PARTITION_MANAGER=0 -D_FORTIFY_SOURCE=2 -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/posix/inf_clock -I/workdir/zephyr/boards/posix/native_posix -I/workdir/nrf/include -Os -imacros /workdir/build-hello/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -imacros /workdir/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/project_test=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -include /workdir/zephyr/arch/posix/include/posix_cheats.h -fno-freestanding -std=c11 -Werror=unused-variable -MD -MT CMakeFiles/app.dir/main.c.obj -MF CMakeFiles/app.dir/main.c.obj.d -o CMakeFiles/app.dir/main.c.obj -c /workdir/project_test/main.c
/workdir/project_test/main.c:10:12: error: ‘unused_var’ defined but not used [-Werror=unused-variable]
   10 | static int unused_var;
      |            ^~~~~~~~~~
cc1: some warnings being treated as errors
```

If this sounds reasonable I'm happy to open a PR with this change, but wanted to flight it here first as an issue to confirm this isn't the intended behavior first.

--- a/cmake/modules/extensions.cmake
+++ b/cmake/modules/extensions.cmake
@@ -500,13 +500,11 @@ function(zephyr_library_compile_options item)
   string(MD5 uniqueness ${item})
   set(lib_name options_interface_lib_${uniqueness})
 
-  if (TARGET ${lib_name})
-    # ${item} already added, ignoring duplicate just like CMake does
-    return()
+  if (NOT TARGET ${lib_name})
+    # Create the unique target only if it doesn't exist.
+    add_library(           ${lib_name} INTERFACE)
+    target_compile_options(${lib_name} INTERFACE ${item} ${ARGN})
   endif()
-
-  add_library(           ${lib_name} INTERFACE)
-  target_compile_options(${lib_name} INTERFACE ${item} ${ARGN})
 
   target_link_libraries(${ZEPHYR_CURRENT_LIBRARY} PRIVATE ${lib_name})
 endfunction()

